From 0e612d3c1b46c16b8bb775fccfd60eb775d54d74 Mon Sep 17 00:00:00 2001
From: Philippe Gerum <rpm@xenomai.org>
Date: Fri, 26 Aug 2022 19:17:42 +0200
Subject: [PATCH 738/985] evl/mutex: fix race in release path

The issue is with the mutex owner handle transitioning to
EVL_NO_HANDLE before transfer_ownership() may switch it to the handle
of the first contender in line. If evl_trylock_mutex() or a user
context manage to slip in with a succcessful cmpxchg() in this tiny
window, two threads may actually end up holding the same lock.

To address this bug, set the handle directly to its final value as a
result of transferring ownership, without transitioning via
EVL_NO_HANDLE, closing the race window.

Signed-off-by: Philippe Gerum <rpm@xenomai.org>
Reported-by: Giulio Moro <giulio@bela.io>
---
 kernel/evl/mutex.c | 12 +++++++-----
 1 file changed, 7 insertions(+), 5 deletions(-)

diff --git a/kernel/evl/mutex.c b/kernel/evl/mutex.c
index e0cf1ba21..d96f8c9f0 100644
--- a/kernel/evl/mutex.c
+++ b/kernel/evl/mutex.c
@@ -899,14 +899,16 @@ void __evl_unlock_mutex(struct evl_mutex *mutex)
 	if (mutex->flags & EVL_MUTEX_CEILING)
 		clear_boost_locked(mutex, curr, EVL_MUTEX_CEILING);
 
+	/*
+	 * The whole logic is based on the invariant that the current
+	 * thread does own the mutex being released.
+	 * IOW: currh == atomic_read(lockp) & ~(EVL_MUTEX_FLCLAIM|EVL_MUTEX_FLCEIL).
+	 */
 	h = atomic_read(lockp);
-	h = atomic_cmpxchg(lockp, h, EVL_NO_HANDLE);
-	if ((h & ~EVL_MUTEX_FLCEIL) != currh) {
-		/* FLCLAIM set, mutex is contended. */
+	if (h & EVL_MUTEX_FLCLAIM) { /* Is there a contender? */
 		transfer_ownership(mutex, curr);
 	} else {
-		if (h != currh)	/* FLCEIL set, FLCLAIM clear. */
-			atomic_set(lockp, EVL_NO_HANDLE);
+		atomic_set(lockp, EVL_NO_HANDLE);
 		untrack_owner(mutex);
 	}
 
-- 
2.30.2

